### PR TITLE
Add Scene_Settings - Config 3 of N

### LINKED ISSUE #666: Persistent storage for global configuration
This one superseds #53

The game browser would make a bit more sense when you could browse (or auto scan as planned for Android) your hard disk and add game directories.
The game directories must be remembered.

This needs changes to the FileFinder:
- [x] CreateDirectory function
- [ ] GetConfigDirectory function
- [ ] GetConfigFile function

CreateDirectory: As the name says

GetConfigDirectory: Create path to config directory if non existant and return a DirectoryTree handle to it. SDL2 has a function for this (can't remember the name). The code could be used as a reference because I don't want a SDL2 dependency.

GetConfigFile: Return path to config file in config directory. Format to be decided, probably INI or JSON (picojson header)

Could probably also store other things in this file like button mappings (joystick), but that's not subject of this issue

``` ini
[GameDirectories]
LastDirectory=
Folder1=
Folder2=
...
Folder9=
```

``` json
{
   "last_directory": "",
   "game_directories" : ["folder1", "folder2"]
}
```

### LINKED ISSUE #713: Configurable game controller/joystick/gamepad buttons
RPG_RT supports joystick/gamepad gameplay. Player also supports this feature.
However, it could be nice to support configurable mappings to customize button mappings, i.e. add lesser common used buttons, e.g. Shift or key numbers to unused game controller buttons.

This issue relies on #666 configuration. Per-game mappings (if implemented) could be saved into a global configuration section to reduce folder writing permissions issues on some platforms.

### LINKED ISSUE #794: Allow more fine-grained control of audio
I guess adding some volume control is currently not needed (Scene_Options anyone? :grinning:), but I think it would be a good idea to allow disabling of music and sound effects separately.
For example I like playing games with only sound effects and having my own music in background.

general_volume

### LINKED ISSUE #1134: Removing Letterboxing in Landscape
I love playing with EasyRPG, and kudos to all those involved.

I'm playing with the surface in Android, and I can't seem to figure out how to stretch the game's screen (variable aspect ratio) to the width of the device in landscape. Since the buttons are overlaid on the same RelativeLayout, I cannot extend the layout itself.

I came across this a few SDL APIs, but it's out of my expertise to pass the device's screen width and height to something like SDL_RenderSetLogicalSize or SDL_RenderSetScale.

Here is a example in practice:
https://www.gamedev.net/resources/_/technical/apis-and-tools/stretching-your-game-to-fit-the-screen-without-letterboxing-sdl2-r3547

Is this possible without lots of headache?

### LINKED ISSUE #1891: Game Settings - Using the 2k3 OpenVideoOptions command
Now that we have the async framework, it is possible to implement the 2k3 OpenVideoOptions command if we think it makes sense on some platforms.

This could async suspend the current game, draw up a scene with video options, then resume the game back from where it started.

optionvideo

### LINKED ISSUE #2721: 2k/3 defaults to --window, but player does not
Maybe @elsemieni can guide us better, but launching a game through a launcher app, the game came prepared for window mode (960x720), but the game launch the game in fullscreen.

I've could fix this with --window, that it was the goal.

All this happened under windows 10 x64.

Thanks in advance.

### LINKED ISSUE #2843: More video options.
As in title. Resolutions such as:
320x240 (the default one I believe), 640x480, 960x720 1280x960 and same for fullscreen (maybe?)

### LINKED ISSUE #2844: [Feature request] Sound volume options.
Either a visual UI element represented as a slider, or command line option will work as long as it saves its state.
And if possible to make separate options for music and FX - would be really cool.

### LINKED ISSUE #2883: Get Mouse Position incorrect when screen zoom is != 1
I'm not sure if it was reported/is a known issue, or if it's an issue with my config, but even without being in fullscreen Get Mouse Position coordinates are not set correctly if I'm not playing in window mode with a Window Zoom of 1 (screen with Window Zoom of 2 for instance) - the base resolution area where the coordinates are taken seems to not be adjusted. (reported by Carbonara :musical_note: , thanks!)

-------------

Remark: This is broken since support for better scaling modes like Bilinear is available as we manually calculate the screen size.

This should also take the new "stretch mode" from the settings into account.

grafik

### PR DESCRIPTION
**Pull request by Ghabry**
Fix #666
Fix #713
Fix #794
Fix #1134
Fix #1891
Fix #2721
Fix #2843
Fix #2844
Fix #2883

----------

This is a WIP to add a settings menu. Early feedback is welcome on the general layout and feature set this should have.

Input:
- [x] Add input remapping for keyboard
- [x] Add input remapping for gamepads

Audio:
- [x] BGM Volume
- [x] SFX Volume

_Not planned_:
- [ ] Configure Midi backend to use
- [ ] Configure Midi soundfont to use
- [ ] Configure which system audio device to output to.
- [ ] Number of output channels
- [ ] Output sample rate
- [ ] Output bit depth
- [ ] Bit-perfect output toggle?

Video:
- [x] Show which renderer is active and leave future support for multiple renderers
- [x] Fullscreen
- [ ] Screen Resolution
- [x] Vsync
- [x] Frame Limiter
- [x] Show FPS

Access:
- [x] Game Browser
- [x] Title Screen
- [x] Menu Scene
- [x] OpenVideoOption event command
- [x] Add a key to open the settings menu

Persistence:
- [x] Load settings to user config file
- [x] Save settings

Others:
- [ ] Refresh UI if video options changed by key press
- [x] Show title screen in background

### CODE REVIEW

**Comment by lychees** [comment]
FYI, there is a setting menu in the yumebooru's multiplayer mode which already include some features (i.e. SFX Volume).
https://github.com/CataractJustice/ynoclient/blob/master/src/game_multiplayer_settings_scene.cpp

**Comment by Ghabry** [comment]
Jenkins: Test this please <3

**Comment by fdelapena** [comment]
So far so good :+1: . In games without title scene, is it showing the default pixel background color? What about games skipping title scene?

Fullscreen does not toggle on Emscripten yet, likely expected (disabling it on emscripten for now might be worth).

Regarding the checklist, output channels is relative to single sample output channels? (16 unique samples, 4 exclusive same-sample channel, total up to 64)  or midi poliphony? (fluidsynth default is 256) or just global output channels meaning mono (1) / stereo (2) / surround (n)?.

Additional ideas, for this or for later (config 4 of N), e.g. toggle pause on focus lost and other:
Click here: https://github.com/EasyRPG/Player/labels/Settings

**Comment by jetrotal** [comment]
Hey guys, silly suggestion:
what about making the look of it like this:
image

It's a mockup based on the game select window:
image

I just saw the Yumebooru multiplayer menu and it's closer to in-game menus as in my suggestion.
image

**Comment by Ghabry** [comment]
I actually thought about the opposite: Moving the options from the game browser in a single top row to win more horizontal space for text (game names in the list) 😅.

**Comment by fdelapena** [comment]
Aligning values to the right as in Yumebooru modification keeps the content cleaner and more readable for both keys and values. The selectable line cursor helps reading the key-value association when the separation is significant.

**Comment by Ghabry** [comment]
I decided to not implement most of the planned functionality by fmatthew because this is too much work to do.

My remaining todo list:

- [ ] Proper storage location for all the stuff (emscripten needs a new mount here to ensure it is remembered on the entire domain)
- [ ] Numeric inputs should use the number input window
- [ ] Frame limiter support
- [ ] Vsync ON/OFF (Restart required)
- [ ] Audio: Global BGM & SE Volume
- [ ] Input: Swapping Primary & Secondary Analog Stick (for Lefties)
- [ ] Input: Swapping AX with BY (for Nintendo fans)

**Comment by TrunXGIT** [comment]
Regarding the Video Options:
In Windowed Mode it would be great if you could choose between all integer scalings (x1,x2,x3,x4,x5...) up to the biggest possible that fits inside the display resolution.
In Fullscreen it might be enough to give the option between the biggest possible integer scaling and one that fills the (vertical) screen space completely even if that means uneven/blurred pixels.
The icing on the cake would be an option for shader effects like CRT/LCD ones. But might be out of scope.

**Comment by aleck099** [comment]
@Ghabry How to modify global audio volume?

**Comment by Mimigris** [comment]
> Ghabry How to modify global audio volume?

There does not seems to have currently a global audio volume, just BGM or SFX, so you have to manually edit both.

> This is a WIP to add a settings menu. Early feedback is welcome on the general layout and feature set this should have.

Was thinking, maybe adding an option to choose a translation for Easy RPG Menu and settings + translation to load if it exist for a game could be added to the menu ? (and if this is something that can't be edited while in-game exceptions for options could be added)

**Comment by aleck099** [comment]
@Mimigris I saw when playing bgm there is a volume parameter, but how can I modify its volume when the bgm is playing?

**Comment by Mimigris** [comment]
> @Mimigris I saw when playing bgm there is a volume parameter, but how can I modify its volume when the bgm is playing?

Not sure if you are really talking here about the setting implementation planned for Easy RPG Player or just talking in terms of coding for a RPG Maker game, so:
Easy RPG Player: Not currently available directly as a build, but this should be available in the settings menu of the Player.
RPG Maker: You have to manually edit each song that play to edit the volume.

If all you want is making your game being less loud when playing you can still change the volume setting of the app in the Volume Mixer of your OS I guess.

**Comment by Ghabry** [comment]
Well, imo this is "feature complete" now on Linux if you want to test.

Have to unbreak the build now for the other platforms and provide a config directory everywhere.

**Comment by Ghabry** [comment]
Suggested input UI by Jetrotal:

unknown

**Comment by jetrotal** [comment]
That suggestions requires always having a system2 hand-icon png.

-----------------------

I'd also suggest limitting the inputs quantity or testing single page per command, 
for cases when the inputs are all huge text chunks like the `JOY_STICK_PRIMARY_UP`

image

**Comment by Ghabry** [comment]
Not as colorful as the suggested one but at least fits on one screen ;)

screenshot_13

-------------

Also pretty close to finishing this (finally!). Only missing is adopting certain homebrew port features to the config scene (stretch, touch ui).

**Comment by Mimigris** [comment]
Just found an issue while testing this too:
Opening the settings menu (F1/assigned key - Open Video Options command) while the screen should be hidden will make the screen be shown to the player once said menu has been closed, while this should not be the case. Seems to be a leftover from opening the default menu.

**Comment by Ghabry** [comment]
@aleck099

In this PR is a new API on the Audio interface ``BGM_SetGlobalVolume``. This writes in the configuration so it is preserved across sessions.

Right now there is no supported way to do it.

**Comment by Ghabry** [comment]
(Not commited yet)

Added a backlight control setting for the 3DS. Yes. There is already one on the touchscreen. They are kept in sync and are saved in the configuration file.

Not tested yet as citra does not emulate this (???).

a

Custom names for the keys are now implemented:

b

c

**Comment by Ghabry** [comment]
@Mimigris 

Thanks. Screen fade-in is resolved. Missed that edge-case :)

------------

Added the emergency reset sequence (I wanted this option anyway. As I want to prevent the typical RetroArch situation where the Ui becomes unusable due to a button mapping error). It is activated by holding 4 arbitrary keys at once. Is displayed in the input remapping as a hint.

There are still ways to make the Player uncontrollable (by leaving the settings while the input is broken) but this is already close to idiotic user error and you can fix it by (somehow) reentering the settings and doing the reset sequence.

**Comment by Ghabry** [comment]
Jenkins: Test this please

**Comment by Ghabry** [comment]
First test results:

-  if you swap [AB], it'll swap them then quit the menu (probably because it then consider the "new" input right after), not a major thing but a funny quirk
- The window size is not preserved right now (maybe add this?)
- vsync not saved (?)
- If you put garbage on config.ini none of the keys works
- submenu for key mapping should be a good idea: game, developer/test, other

**Comment by Ghabry** [comment]
Suggested by @jetrotal . This was simple to add

screenshot_3

**Comment by aleck099** [comment]
@Ghabry That's great

**Comment by Ghabry** [comment]
Geting close. Just few bug fixes and platform tests missing.

Getting close.
Redesigned the Game Browser to ensure all menu items fit. You will get used to it.

screenshot_1

New settings category "Engine" that contains a few entries that fit nowhere else:

screenshot_2

**Comment by Ghabry** [comment]
Except for checking some ports this is now feature complete and ready for review. :tada: This took over a year to finish with multiple long breaks because this was so frustrating to program.

Sorry for so many LoC changes and 50+ commits. There was no existing code in the Player to handle persistent configurations so everything had to be written from scratch.

At least further additions of config options will be much less code...

**Review by fdelapena — APPROVED** [review]

**Comment by Ghabry** [comment]
Had to retest Vita and Switch to solve some final bugs.

Only pending tests are Android and a retest on Windows

**Comment by Mimigris** [comment]
> Only pending tests are Android and a retest on Windows

The only issue I've found on my side while re-testing the latest version on my Windows is that the cursor while in the Settings menu part where you can remap your keys "Input -> Key/Button mapping" and then _Game / Engine / Developer_ is not properly tracked like the other menus (If your cursor is on the 22th item in the Engine part, when opening the Developer part your cursor will also be on the 22th slot although it is not possible normally).
screenshot_1
screenshot_0

**Comment by Ghabry** [comment]
Also finished Android now.

@Mimigris your reported bug should be resolved

File: src/platform/android/org_easyrpg_player_player_EasyRpgPlayerActivity.cpp [hunk 1]
```diff
@@ -2,41 +2,37 @@
  * This file is part of EasyRPG Player
  *
  * Copyright (c) 2017 EasyRPG Project. All rights reserved.
```
Comment (fdelapena) [inline src/platform/android/org_easyrpg_player_player_EasyRpgPlayerActivity.cpp:4]: At least this header still has the © year, a good chance to remove it

Comment (Ghabry) [reply src/platform/android/org_easyrpg_player_player_EasyRpgPlayerActivity.cpp:4]: and not touched since 6 years xD

**Pull request merged**